File: lib/features/subscription/domain/subscription_status.dart
import 'package:cloud_firestore/cloud_firestore.dart';

enum SubscriptionState { active, trialing, expired, canceled }

/// Mirrors `users/{uid}/subscription/status`. This is the single source of
/// truth for what plan a user has and what a future billing provider
/// (Stripe/RevenueCat/Play Billing/Lemon Squeezy/Paddle) would keep in sync
/// via webhooks.
class SubscriptionStatus {
  const SubscriptionStatus({
    required this.planId,
    required this.status,
    required this.startedAt,
    this.expiresAt,
    required this.billingProvider,
    required this.isTrialActive,
    this.trialEndsAt,
  });

  final String planId;
  final SubscriptionState status;
  final DateTime startedAt;
  final DateTime? expiresAt;

  /// 'none' until a real provider is wired up, then e.g. 'stripe',
  /// 'revenuecat', 'play_billing', 'lemon_squeezy', 'paddle'.
  final String billingProvider;
  final bool isTrialActive;
  final DateTime? trialEndsAt;

  factory SubscriptionStatus.fromFirestore(Map<String, dynamic> data) {
    return SubscriptionStatus(
      planId: data['planId'] as String? ?? 'starter',
      status: SubscriptionState.values.firstWhere(
        (s) => s.name == data['status'],
        orElse: () => SubscriptionState.active,
      ),
      startedAt: (data['startedAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
      expiresAt: (data['expiresAt'] as Timestamp?)?.toDate(),
      billingProvider: data['billingProvider'] as String? ?? 'none',
      isTrialActive: data['isTrialActive'] as bool? ?? false,
      trialEndsAt: (data['trialEndsAt'] as Timestamp?)?.toDate(),
    );
  }

  Map<String, dynamic> toFirestore() {
    return {
      'planId': planId,
      'status': status.name,
      'startedAt': Timestamp.fromDate(startedAt),
      'expiresAt': expiresAt == null ? null : Timestamp.fromDate(expiresAt!),
      'billingProvider': billingProvider,
      'isTrialActive': isTrialActive,
      'trialEndsAt': trialEndsAt == null ? null : Timestamp.fromDate(trialEndsAt!),
    };
  }
}

File: steady_progress/src/features/subscription/subscription-status.ts
import { dateFromFirestore } from '../../core/data/firestore-values.ts';

export type SubscriptionState = 'active' | 'trialing' | 'expired' | 'canceled';
export type SubscriptionStatus = {
  planId: string;
  status: SubscriptionState;
  startedAt: Date | null;
  expiresAt: Date | null;
  billingProvider: string;
  isTrialActive: boolean;
  trialEndsAt: Date | null;
};

export function subscriptionStatusFromDocument(data: Record<string, unknown>): SubscriptionStatus {
  const state = data.status;
  return {
    planId: typeof data.planId === 'string' ? data.planId : 'starter',
    status: state === 'trialing' || state === 'expired' || state === 'canceled' ? state : 'active',
    startedAt: dateFromFirestore(data.startedAt),
    expiresAt: dateFromFirestore(data.expiresAt),
    billingProvider: typeof data.billingProvider === 'string' ? data.billingProvider : 'none',
    isTrialActive: data.isTrialActive === true,
    trialEndsAt: dateFromFirestore(data.trialEndsAt),
  };
}

export function isSubscriptionEntitled(status: SubscriptionStatus, now = new Date()) {
  if (status.status === 'expired' || status.status === 'canceled') return false;
  if (status.status === 'trialing' && status.trialEndsAt && now >= status.trialEndsAt) return false;
  return !status.expiresAt || now < status.expiresAt;
}
